src/io.cpp

```diff
@@ -102,38 +116,51 @@ void io_buffer_t::begin_filling(autoclose_fd_t fd) {
 
     // Run our function to read until the receiver is closed.
     // It's OK to capture 'this' by value because 'this' waits for the promise in its dtor.
-    fd_monitor_item_t item;
-    item.fd = std::move(fd);
-    item.callback = [this, promise](autoclose_fd_t &fd, item_wake_reason_t reason) {
-        ASSERT_IS_BACKGROUND_THREAD();
-        // Only check the shutdown flag if we timed out or were poked.
-        // It's important that if select() indicated we were readable, that we call select() again
-        // allowing it to time out. Note the typical case is that the fd will be closed, in which
-        // case select will return immediately.
-        bool done = false;
-        if (reason == item_wake_reason_t::readable) {
-            // select() reported us as readable; read a bit.
-            auto buffer = buffer_.acquire();
-            ssize_t ret = read_once(fd.fd(), buffer);
-            done = (ret == 0 || (ret < 0 && errno != EAGAIN && errno != EWOULDBLOCK));
-        } else if (shutdown_fillthread_) {
-            // Here our caller asked us to shut down; read while we keep getting data.
-            // This will stop when the fd is closed or if we get EAGAIN.
-            auto buffer = buffer_.acquire();
-            ssize_t ret;
-            do {
-                ret = read_once(fd.fd(), buffer);
-            } while (ret > 0);
-            done = true;
-        }
-        if (done) {
-            fd.close();
-            promise->set_value();
-        }
-    };
-    this->item_id_ = fd_monitor().add(std::move(item));
+    auto args = new callback_args_t;
+    args->instance = this;
+    args->promise = std::move(promise);
+
+    item_id_ =
+        fd_monitor().add_item(fd.acquire(), kNoTimeout, (::c_void *)item_callback_trampoline, (::c_void *)args);
 }
 
+/// This is a hack to work around the difficulties in passing a capturing lambda across FFI
+/// boundaries. A static function that takes a generic/untyped callback parameter is easy to
+/// marshall with the basic C ABI.
+void io_buffer_t::item_callback(autoclose_fd_t2 &fd, uint8_t r, callback_args_t *args) {
+    item_wake_reason_t reason = (item_wake_reason_t)r;
+    auto &promise = *args->promise;
```

Comment:
we might want to port ASSERT_IS_BACKGROUND_THREAD and friends but that's probably not important

Reply:
There’s no need since now it’s in a separate type that’s instantiated on the background thread. It used to be a member function of the same object shared on both foreground and background.